python/graphframes/examples/belief_propagation.py

```diff
@@ -0,0 +1,173 @@
+                # receive messages and update beliefs for vertices of the current color
+                newBeliefCol = sqlfunctions.when(
+                    (v['color'] == color) & (aggregates['aggMess'].isNotNull()),
+                    logistic(aggregates['aggMess'] + v['a'])
+                ).otherwise(v['belief'])  # keep old beliefs for other colors
+                newVertices = (v
+                    .join(aggregates, on=(v['id'] == aggregates['id']), how='left_outer')
+                    .drop(aggregates['id'])  # drop duplicate ID column (from outer join)
+                    .withColumn('newBelief', newBeliefCol)  # compute new beliefs
+                    .drop('aggMess')  # drop messages
+                    .drop('belief')  # drop old beliefs
+                    .withColumnRenamed('newBelief', 'belief')
+                )
+                # cache new vertices using workaround for SPARK-1334
+                cachedNewVertices = AM.getCachedDataFrame(newVertices)
+                gx = GraphFrame(cachedNewVertices, gx.edges)
+
+        # Drop the "color" column from vertices
+        return GraphFrame(gx.vertices.drop('color'), gx.edges)
+
+    @staticmethod
+    def _colorGraph(g):
+        """Given a GraphFrame, choose colors for each vertex.
+
+        No neighboring vertices will share the same color. The number of colors is minimized.
+
+        This is written specifically for grid graphs. For non-grid graphs, it should be generalized,
+        such as by using a greedy coloring scheme.
+
+        :param g: Grid graph generated by :meth:`Graphs.gridIsingModel()`
+        :return: Same graph, but with a new vertex column "color" of type Int (0 or 1)
+
+        """
+
+        colorUDF = sqlfunctions.udf(lambda i, j: (i + j) % 2, returnType=types.IntegerType())
+        v = g.vertices.withColumn('color', colorUDF(sqlfunctions.col('i'), sqlfunctions.col('j')))
+        return GraphFrame(v, g.edges)
+
+    @staticmethod
+    def _sigmoid(x):
+        """Numerically stable sigmoid function 1 / (1 + exp(-x))"""
+        if not x:
+            return None
+        if x >= 0:
+            z = math.exp(-x)
+            return 1 / (1 + z)
+        else:
+            z = math.exp(x)
+            return z / (1 + z)
+
+
+def main():
+    """Run the belief propagation algorithm for an example problem."""
+    # setup context
+    conf = SparkConf().setAppName("BeliefPropagation example")
+    sc = SparkContext.getOrCreate(conf)
+    sql = SQLContext.getOrCreate(sc)
+
+    # create graphical model g of size 3 x 3
+    g = Graphs(sql).gridIsingModel(3)
+    print("Original Ising model:")
+    g.vertices.show()
+    g.edges.show()
+
+    # run BP for 5 iterations
+    numIter = 5
+    results = BeliefPropagation.runBPwithGraphFrames(g, numIter)
+
+    # display beliefs
+    beliefs = results.vertices.select('id', 'belief')
+    print("Done with BP. Final beliefs after {} iterations:".format(numIter))
+    beliefs.show()
```

Comment:
I tested the BeliefPropagation function by submitting similar identical manually created graph to both Python and Spark implementations and got identical results.

Also I noticed this comment in the Scala API docs of `org.apache.spark.sql.functions.randn`
```
Note    This is indeterministic when data partitions are not fixed.
```
I suppose this is the reason for getting different random numbers.